src/server/game/Scripting/ScriptMgr.cpp

```diff
@@ -1599,7 +1600,19 @@ CreatureAI* ScriptMgr::GetCreatureAI(Creature* creature)
     ASSERT(creature);
 
     GET_SCRIPT_RET(CreatureScript, creature->GetScriptId(), tmpscript, nullptr);
-    return tmpscript->GetAI(creature);
+    CreatureAI* ai = tmpscript->GetAI(creature);
+
+    // Don't allow PetAI if there's no charm info in the creature
+    if (dynamic_cast<PetAI*>(ai))
```

Comment:
"meh"

can't we make a new virtual method on `CreatureAI` that's called `IsValid` or something?

Reply:
no, because then you are putting control back in hands of script and the point of this validation is to harden the core against bad scripts

Reply:
@Shauren what if we do
```c++
virtual bool CreatureAI::IsAIAllowed(Creature* creature) { return true;}
```
and then
```c++
bool PetAI::IsAIAllowed(Creature* creature) final override { return !!creature->GetCharmInfo(); }
```
using final override we should keep the check in the core without allowing any script to override it

Reply:
@Shauren but a script has to _active_ suppress an inherited `IsAIAllowed`, while currently just inheriting from `PetAI` _passively_ causes crashes.

Reply:
Then set PetAi as final so no one can inherit it

Reply:
that'd lead to tons of code duplication

Reply:
how about throwing an exception from PetAI constructor when charminfo is missing?

Reply:
and catching that in `ScriptMgr`? that's reasonable, :+1:

Reply:
if we catch it there we'd be leaking the half-constructed object, no?

Reply:
no, thats not how exceptions from constructors work, you do not return anything, you dont even get a pointer to use with operator delete (tl;dr, dont worry about it)